src/drivers/intel/ssp/ssp.c

```diff
@@ -712,6 +712,63 @@ static int ssp_set_config(struct dai *dai, struct ipc_config_dai *common_config,
 	ssp->state[DAI_DIR_PLAYBACK] = COMP_STATE_PREPARE;
 	ssp->state[DAI_DIR_CAPTURE] = COMP_STATE_PREPARE;
 
+	switch (config->flags & SOF_DAI_CONFIG_FLAGS_MASK) {
+	case SOF_DAI_CONFIG_FLAGS_HW_PARAMS:
+		if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_ES) {
+			ret = ssp_mclk_prepare_enable(dai);
+			if (ret < 0)
+				goto out;
+
+			ssp->clk_active |= SSP_CLK_MCLK_ES_REQ;
+
+			dai_info(dai, "ssp_set_config(): hw_params stage: enabled MCLK clocks for SSP%d...",
+				 dai->index);
+		}
+
+		if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES) {
+			bool enable_sse = false;
+
+			if (!(ssp->clk_active & SSP_CLK_BCLK_ACTIVE))
+				enable_sse = true;
+
+			ret = ssp_bclk_prepare_enable(dai);
+			if (ret < 0)
+				goto out;
+
+			ssp->clk_active |= SSP_CLK_BCLK_ES_REQ;
+
+			if (enable_sse) {
+				/* enable port */
+				ssp_update_bits(dai, SSCR0, SSCR0_SSE, SSCR0_SSE);
+
+				dai_info(dai, "ssp_set_config(): SSE set for SSP%d", dai->index);
+			}
+
+			dai_info(dai, "ssp_set_config(): hw_params stage: enabled BCLK clocks for SSP%d...",
+				 dai->index);
+		}
+		break;
+	case SOF_DAI_CONFIG_FLAGS_HW_FREE:
+		if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES) {
+			dai_info(dai, "ssp_set_config(): hw_free stage: releasing BCLK clocks for SSP%d...",
+				 dai->index);
+			if (ssp->clk_active & SSP_CLK_BCLK_ACTIVE) {
+				ssp_update_bits(dai, SSCR0, SSCR0_SSE, 0);
```

Comment:
I think we need to consider the streaming state of both TX&RX before disabling the port. like in ssp_stop() function.

Reply:
@brentlu Yes we did it at line 272
```
	/* is playback/capture already running */
	if (ssp->state[DAI_DIR_PLAYBACK] == COMP_STATE_ACTIVE ||
	    ssp->state[DAI_DIR_CAPTURE] == COMP_STATE_ACTIVE) {
		dai_info(dai, "ssp_set_config(): playback/capture active. Ignore config");
		goto out;
	}
```

Reply:
thanks for explanation.